pkg/gpu/cuda/fatbin.go

```diff
@@ -0,0 +1,254 @@
+	ConstantMem uint64                          // Size of the constant memory used by the kernel
+}
+
+// GetKernel returns the kernel with the given name and SM version from the fatbin
+func (fb *Fatbin) GetKernel(name string, smVersion uint32) *CubinKernel {
+	key := CubinKernelKey{Name: name, SmVersion: smVersion}
+	if _, ok := fb.Kernels[key]; !ok {
+		return nil
+	}
+	return fb.Kernels[key]
+}
+
+type fatbinHeader struct {
+	Magic      uint32
+	Version    uint16
+	HeaderSize uint16
+	FatSize    uint64 // not including the header
+}
+
+func (fbh *fatbinHeader) validate() error {
+	if fbh.Magic != fatbinMagic {
+		return fmt.Errorf("invalid fatbin header, magic number %x does not match expected %x", fbh.Magic, fatbinMagic)
+	}
+	if fbh.Version != fatbinHeaderVersion {
+		return fmt.Errorf("invalid fatbin header, version %d does not match expected %d", fbh.Version, fatbinHeaderVersion)
+	}
+
+	return nil
+}
+
+type fatbinData struct {
+	Kind                    uint16
+	Version                 uint16
+	HeaderSize              uint32
+	PaddedPayloadSize       uint32
+	Unknown0                uint32 // check if it's written into separately
+	PayloadSize             uint32
+	Unknown1                uint32
+	Unknown2                uint32
+	SmVersion               uint32
+	BitWidth                uint32
+	Unknown3                uint32
+	Unknown4                uint64
+	Unknown5                uint64
+	UncompressedPayloadSize uint64
+}
+
+func (fbd *fatbinData) dataKind() fatbinDataKind {
+	return fatbinDataKind(fbd.Kind)
+}
+
+func (fbd *fatbinData) validate() error {
+	dataKind := fbd.dataKind()
+	if dataKind < fatbinDataMinKind || dataKind > fatbinDataMaxKind {
+		return fmt.Errorf("kind %d is not in the expected range [%d, %d]", dataKind, fatbinDataMinKind, fatbinDataMaxKind)
+	}
+	if fbd.Version != fatbinDataVersion {
+		return fmt.Errorf("version %d does not match expected %d", fbd.Version, fatbinDataVersion)
+	}
+
+	return nil
+}
+
+// ParseFatbinFromELFFilePath opens the given path and parses the resulting ELF for CUDA kernels
+func ParseFatbinFromELFFilePath(path string) (*Fatbin, error) {
+	elfFile, err := elf.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open ELF file %s: %w", path, err)
+	}
+	defer elfFile.Close()
+
+	return ParseFatbinFromELFFile(elfFile)
+}
+
+func getBufferOffset(buf io.Seeker) int64 {
+	offset, _ := buf.Seek(0, io.SeekCurrent)
+	return offset
+}
+
+// ParseFatbinFromPath parses the fatbin sections of the given ELF file and returns the information found in it
+func ParseFatbinFromELFFile(elfFile *elf.File) (*Fatbin, error) {
+	fatbin := &Fatbin{
+		Kernels: make(map[CubinKernelKey]*CubinKernel),
+	}
+
+	for _, sect := range elfFile.Sections {
+		// CUDA embeds the fatbin data in sections named .nv_fatbin or __nv_relfatbin
+		if sect.Name != ".nv_fatbin" && sect.Name != "__nv_relfatbin" {
+			continue
+		}
+
+		data, err := sect.Data()
```

Comment:
have you considered using section.Open instead of Data(), it is using  a special io.SectionReader under the hood and also allows seek operations. 
I also think it can handle the SHT_NOBITS internally.

that way you don't need to allocate the buffer in memory for the entire section.